detection_rules/rule.py

```diff
@@ -255,9 +255,9 @@ class ThresholdCardinality:
             field: str
             value: definitions.ThresholdValue
 
-        field: List[definitions.NonEmptyStr]
+        field: definitions.StringListMaxSizeThree
         value: definitions.ThresholdValue
-        cardinality: Optional[ThresholdCardinality]
+        cardinality: Optional[List[ThresholdCardinality]]
```

Comment:
We need to generate new schema JSON, and might need to manually update our old ones.
I believe this is 7.13+ specifically

Reply:
Looks like there is a pre-existing bug in `update-schemas` command

<details>

```
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/detection_rules/__main__.py", line 34, in <module>
    main()
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/detection_rules/__main__.py", line 31, in main
    root(prog_name="detection_rules")
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/venv38/lib/python3.8/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/venv38/lib/python3.8/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/venv38/lib/python3.8/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/venv38/lib/python3.8/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/venv38/lib/python3.8/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/venv38/lib/python3.8/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/detection_rules/devtools.py", line 511, in update_schemas
    cls.save_schema()
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/detection_rules/rule.py", line 187, in save_schema
    schema = cls.jsonschema()
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/detection_rules/utils.py", line 241, in wrapped
    _cache[func_key][cache_key] = f(*args, **kwargs)
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/detection_rules/mixins.py", line 90, in jsonschema
    jsonschema = marshmallow_jsonschema.JSONSchema().dump(c)
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/venv38/lib/python3.8/site-packages/marshmallow_jsonschema/base.py", line 328, in dump
    return super().dump(obj, **kwargs)
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/venv38/lib/python3.8/site-packages/marshmallow/schema.py", line 552, in dump
    result = self._serialize(processed_obj, many=many)
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/venv38/lib/python3.8/site-packages/marshmallow/schema.py", line 520, in _serialize
    value = field_obj.serialize(attr_name, obj, accessor=self.get_attribute)
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/venv38/lib/python3.8/site-packages/marshmallow/fields.py", line 318, in serialize
    return self._serialize(value, attr, obj, **kwargs)
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/venv38/lib/python3.8/site-packages/marshmallow/fields.py", line 1805, in _serialize
    return self._serialize_method(obj)
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/venv38/lib/python3.8/site-packages/marshmallow_jsonschema/base.py", line 155, in get_properties
    schema = self._get_schema_for_field(obj, field)
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/venv38/lib/python3.8/site-packages/marshmallow_jsonschema/base.py", line 256, in _get_schema_for_field
    pytype = self._get_python_type(field)
  File "/Users/jibarra/PycharmProjects/detection-rules-fork/venv38/lib/python3.8/site-packages/marshmallow_jsonschema/base.py", line 241, in _get_python_type
    raise UnsupportedValueError("unsupported field type %s" % field)
marshmallow_jsonschema.exceptions.UnsupportedValueError: unsupported field type <fields.Union(default=<marshmallow.missing>, attribute=None, validate=None, required=True, load_only=False, dump_only=False, missing=<marshmallow.missing>, allow_none=False, error_messages={'required': 'Missing data for required field.', 'null': 'Field may not be null.', 'validator_failed': 'Invalid value.'})>
```

</details>

It is happening on `machine_learning_job_id` for `MachineLearningRuleData`

Reply:
Fixed in #1359 and pulled into this branch. You should be good now.
I think in another PR we should add test that those JSON files match.

Reply:
updated schemas in 9ba3c3b50255a8f1a5ea690e4a1e86cffd136611

Also looks like it was introduced in 7.12 schema. Fix, along with unit test in 9ba3c3b50255a8f1a5ea690e4a1e86cffd136611

Reply:
thoughts on this?
```suggestion
        cardinality: Optional[Union[ThresholdCardinality, List[ThresholdCardinality]]]
```

Reply:
The API schema was never implemented as a non-list; it was just a mistake in the interpretation here, so I don't think we should make that change

Reply:
oh right cardinality. sorry that was ML jobs and I'm mixing things